### PR TITLE
types: add types for remark-lint and lint rules

### PR DESCRIPTION
**Pull request by ChristianMurphy**
:warning: This PR is WIP
:notebook: This PR will be massive, 2 files for each package + 4 test/config files.

### CODE REVIEW

**Comment by wooorm** [comment]
Awesome!

File: tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{
+  "compilerOptions": {
+    "lib": ["es2015"],
+    "strict": true,
+    "noImplicitAny": true,
+    "noImplicitThis": true,
+    "strictNullChecks": true,
+    "strictFunctionTypes": true,
+    "baseUrl": ".",
+    "paths": {
```
Comment (wooorm) [inline tsconfig.json]: these paths could perhaps also be automated btw:

```
var cur = JSON.parse(current)
var next = {...cur, paths: generate()}
var doc = JSON.stringify(next, null, 2) + '\n'
```

Comment (ChristianMurphy) [reply tsconfig.json]: refactored paths to be pattern based, all lint packages should now be matched.

```json
"paths": {
  "remark-*": ["packages/remark-*"]
}
```

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// TypeScript Version: 3.4
+
+export function placeholderForDtslint(): void
```
Comment (Rokt33r) [inline index.d.ts]: Why do we need this file?

Comment (ChristianMurphy) [reply index.d.ts]: dtslint needs an `index.d.ts` file at the root of where it is called.
This exists purely to satisfy that constraint.

If there is an option to not require it, or an elegant want to call `dslint` for each package.
I'd be happy to remove it.

Comment (Rokt33r) [reply index.d.ts]: I see.It seems the problem has been reported in here...!
https://github.com/microsoft/dtslint/issues/213

Comment (Rokt33r) [reply index.d.ts]: I guess we don't have that much choice. It is needed for dtslint to determine supported typescript versions too.

Anyway, everything LGTM. 👍

**Comment by wooorm** [comment]
Closing this as this is stalled. Still, a PR with types for these packages would be accepted!

**Pull request closed without merging**